std/heap.zig

```diff
@@ -89,21 +78,42 @@ pub const DirectAllocator = struct {
 
                 return @intToPtr([*]u8, aligned_addr)[0..n];
             },
-            Os.windows => {
-                const amt = n + alignment + @sizeOf(usize);
-                const optional_heap_handle = @atomicLoad(?HeapHandle, &self.heap_handle, builtin.AtomicOrder.SeqCst);
-                const heap_handle = optional_heap_handle orelse blk: {
-                    const hh = os.windows.HeapCreate(0, amt, 0) orelse return error.OutOfMemory;
-                    const other_hh = @cmpxchgStrong(?HeapHandle, &self.heap_handle, null, hh, builtin.AtomicOrder.SeqCst, builtin.AtomicOrder.SeqCst) orelse break :blk hh;
-                    _ = os.windows.HeapDestroy(hh);
-                    break :blk other_hh.?; // can't be null because of the cmpxchg
-                };
-                const ptr = os.windows.HeapAlloc(heap_handle, 0, amt) orelse return error.OutOfMemory;
-                const root_addr = @ptrToInt(ptr);
-                const adjusted_addr = mem.alignForward(root_addr, alignment);
-                const record_addr = adjusted_addr + n;
-                @intToPtr(*align(1) usize, record_addr).* = root_addr;
-                return @intToPtr([*]u8, adjusted_addr)[0..n];
+            .windows => {
+                const w = os.windows;
+
+                if (alignment <= os.page_size) {
```

Comment:
Below you have written:

>  On Windows, VirtualAlloc returns addresses aligned to a 64K boundary,

I was unable to find the documentation that states this, but if it is true, does that mean we can use `64 * 1024` here rather than `os.page_size`?

Reply:
I don't think it's documented anywhere, but there are a few people mentionning this online. As always, Raymond Chen delivers: https://devblogs.microsoft.com/oldnewthing/?p=42223
Windows being windows, one can possibly rely on this being true, but that's more of an implementation detail actually. I wouldn't mind using it, but I leave the choice up to you.

Reply:
Nice find.

I don't want to rely on undocumented behavior that could possibly change, and I don't think we have to.

Here's my proposal:

 * Without testing for the desired alignment, do the `VirtualAlloc` for the minimal size needed (the code currently in the if() block
 * Check the alignment and see if it's good enough. It's extremely likely to be fine.
    * This check can be optimized because alignment is guaranteed to be a power of 2. I believe that would be: `(addr & (alignment - 1)) == 0`.
 * Otherwise, fall back to the loop for very large alignments.

This will have the benefit of the fast path being fast while taking advantage of any extra alignment that VirtualAlloc gives.